lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,151 @@
+// with ability to enable filesystem notification watch and reopen file on specific events.
+func NewFileSharedWriter(logFileName string, flag int, mode fs.FileMode) (*FileSharedWriter, error) {
+	if logFileName == "" {
+		return nil, trace.BadParameter("log file name is not set")
+	}
+	logFile, err := os.OpenFile(logFileName, flag, mode)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create the log file")
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &FileSharedWriter{
+		logFileName: logFileName,
+		fileFlag:    flag,
+		fileMode:    mode,
+		file:        logFile,
+		watcher:     watcher,
+	}, nil
+}
+
+// Write writes len(b) bytes from b to the File.
+func (s *FileSharedWriter) Write(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.file.Write(b)
+}
+
+// Reopen closes the file and opens it again using APPEND mode.
+func (s *FileSharedWriter) Reopen() (err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if err := s.file.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+	s.file, err = os.OpenFile(s.logFileName, s.fileFlag, s.fileMode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// RunWatcherReopen runs a filesystem watcher for rename/remove events to reopen the log.
+func (s *FileSharedWriter) RunWatcherReopen() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.runWatcherFunc(s.Reopen)
+}
+
+// runWatcherFunc spawns goroutine with the watcher loop to consume events of renaming
+// or removing the log file to trigger the action function when event appeared.
+func (s *FileSharedWriter) runWatcherFunc(action func() error) error {
+	if s.watcher.WatchList() == nil {
+		return trace.BadParameter("watcher is already closed")
+	}
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-s.watcher.Events:
+				if !ok {
+					return
+				}
+				if s.logFileName == event.Name && (event.Has(fsnotify.Rename) || event.Has(fsnotify.Remove)) {
+					slog.DebugContext(context.Background(), "Log file was moved/removed", "file", event.Name)
+					if err := action(); err != nil {
+						slog.ErrorContext(context.Background(), "Failed to take action", "error", err, "file", event.Name)
+						continue
+					}
+				}
+			case err, ok := <-s.watcher.Errors:
+				if !ok {
+					return
+				}
+				slog.ErrorContext(context.Background(), "Error received on logger watcher", "error", err)
+			}
+		}
+	}()
+
+	logDirParent := filepath.Dir(s.logFileName)
+	if err := s.watcher.Add(logDirParent); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// Close stops the internal watcher and close the log file.
+func (s *FileSharedWriter) Close() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if err := s.watcher.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := s.file.Close(); err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
This will leave the file open if the watcher fails to close (and presumably the watcher will forever fail to close cleanly again afterwards, so the file will be open forever). Something like `return trace.NewAggregate(s.watcher.Close(), s.file.Close())` would be better.